### ISSUE TITLE
Is RocketChat not compatible with OAuth2 and Office365?

### ISSUE THREAD

**Issue by basitmohammad** [RocketChat/Rocket.Chat#6809]
I can't set up OAuth with Office365, please help. Is Oauth2 not compatible or supported?

Rocket.Chat Version: 0.54.2
Running Instances: 1
DB Replicaset OpLog: Disabled
Node Version: v4.5.0

**Comment by bluenevus** [comment]
URL: https://login.microsoftonline.com/common
token path: /oauth2/token
identity path: /me
authorize path: /oauth2/authorize
scope: openid
username field: userPrinciplename
and the correct callback url with id and secret configured

same issue.  first it launches, looks like its going to work then it says internal server error

saml for o365 works just can't get oauth to work which we need for outside users

Got this error
oauth.js:431) [35mError in OAuth Server: Failed to fetch identity from office365 at https://login.microsoftonline.com/me. failed [404]
[34mI20170412-19:44:06.805(0) Exception while invoking method 'login' Error: Failed to fetch identity from office365 at https://login.microsoftonline.com/me. failed [404]   at CustomOAuth.getIdentity (/opt/Rocket.Chat/programs/server/packages/rocketchat_custom-oauth.js:198:17)   at Object.handleOauthRequest (/opt/Rocket.Chat/programs/server/packages/rocketchat_custom-oauth.js:214:25)   at OAuth._requestHandlers.(anonymous function) (packages/oauth2.js:27:31)   at middleware (packages/oauth.js:203:5)   at packages/oauth.js:176:5  

now this error

Exception while invoking method 'login' Error: Failed to complete OAuth handshake with oauthoffice365 at https://login.microsoftonline.com/common/oauth2/token. failed [400] {"error":"invalid_grant","error_description":"AADSTS70000: Transmission data parser failure: Authorization Code is malformed or invalid

**Comment by geekgonecrazy** [comment]
I don't believe /me is the proper identity path.  Actually digging through i'm not finding such an equivalent route. 

Microsofts authentication services are a bit of a mess.  They've been doing a bit better in recent years.  

If you can find the identity path in their documentation somewhere you could probably get this to work.  Otherwise I think we're going to have to add one specifically for office365

**Comment by bluenevus** [comment]
Thanks.

 

I tried to use outlook as the identity url to get the identity because that is how I saw some other apps grab the identity but that didn’t work either.  Specifically  <https://outlook.office365.com/api/v1.0/me> https://outlook.office365.com/api/v1.0/me  because it looked similar to your example.

 

-- 

Vr, Frank

**Comment by bluenevus** [comment]
In fact, this is the error when I tried the identity url above

Exception while invoking method 'login' Error: Failed to fetch identity from oauthoffice365 at https://outlook.office365.com/api/v1.0/me. failed [401]   at CustomOAuth.getIdentity (/opt/Rocket.Chat/programs/server/packages/rocketchat_custom-oauth.js:198:17)   at Object.handleOauthRequest (/opt/Rocket.Chat/programs/server/packages/rocketchat_custom-oauth.js:214:25)   at OAuth._requestHandlers.(anonymous function) (packages/oauth2.js:27:31)   at middleware (packages/oauth.js:203:5)   at packages/oauth.js:176:5

**Comment by geekgonecrazy** [comment]
@bluenevus how did you generate your oauth clientid / secret?  Seems to be dozens of different instructions sets floating around.  All of them about as clear as mud.  :grin: I remember at one point having to setup things via Azure AD.  But then I thought there was a more recent set that was almost a wizard via the graph api.

**Comment by bluenevus** [comment]
Go to Microsoft app dev here  <https://apps.dev.microsoft.com/> https://apps.dev.microsoft.com/

 

Register you app much like you would in facebook, twitter or github.  It will generate client id and secret.  Put your call back url there.

 

-- 

Vr, Frank

**Comment by bluenevus** [comment]
its pretty much the common microsoft login whether o365, live etc

**Comment by geekgonecrazy** [comment]
The endpoint seems to be: https://login.microsoftonline.com/common/openid/userinfo

Found here: https://login.microsoftonline.com/common/.well-known/openid-configuration

**Comment by bluenevus** [comment]
Worked like a champ.  We have been struggling for weeks with calls to Microsoft on this.  Thank you, Thank you, Thank you.  You have an amazing product we are trying to pilot

 

-- 

Vr, Frank

**Comment by geekgonecrazy** [comment]
@bluenevus perfect!  What fields did you end up with? That way we can pass the knowledge forward for others :grin:

**Comment by bluenevus** [comment]
For oauth

 

*	url  <https://login.microsoftonline.com/common> https://login.microsoftonline.com/common
*	token path /oauth2/token
*	identity path /openid/userinfo
*	authorize path /oauth2/authorize
*	scope openid
*	token sent via header
*	id from https://apps.dev.microsoft.com
*	secret from  <https://apps.dev.microsoft.com> https://apps.dev.microsoft.com
*	login style popup
*	username field: just doesn’t work.  It should be UserPrincipalName but I get an error that says its not found in data stream.  Tried others like mail and email and it’s the same.  So the user can’t expect it to be the same account if their email is the same registered in say o365 and github…won’t be the same user and won’t allow them to register with the same username
*	merge users really doesn’t matter because it can’t find the username in the data stream

 

-- 

Vr, Frank

**Comment by geekgonecrazy** [comment]
I've managed to duplicate every thing except a successful login :grin: 

Heres what I have on my app page on microsoft. 

![image](https://cloud.githubusercontent.com/assets/51996/25507087/a4f08eee-2b6f-11e7-839e-6ab54e3ac50c.png)

Is there something missing from here?

**Comment by bluenevus** [comment]
Has to be https.  Will not work with http.

 

-- 

Vr, Frank

**Comment by JannikZed** [comment]
Hi,
I always receive the following:
   

>  Error in OAuth Server: Failed to complete OAuth handshake with o365 at https://login.microsoftonline.com/common/oauth2/token. failed [400] {"error":"invalid_request","error_description":"AADSTS90014: The request body must contain the following parameter: 'client_id'.\

**Comment by KoAi** [comment]
Hi,
From my part, with same config mention above I get following error:
> Exception while invoking method 'login' Error: Failed to complete OAuth handshake with o365 at https://login.microsoftonline.com/common/oauth2/token. failed [400]

Did somebody make it works ?

**Comment by bluenevus** [comment]
This is your problem.  
![image](https://user-images.githubusercontent.com/3675043/28471171-3b59a320-6e0a-11e7-90cf-cd294d8e1abc.png)

it has to be https

**Comment by bluenevus** [comment]
![image](https://user-images.githubusercontent.com/3675043/28471212-60effb16-6e0a-11e7-9e54-55c1dba6e19e.png)

**Comment by bluenevus** [comment]
![image](https://user-images.githubusercontent.com/3675043/28471220-69bbb640-6e0a-11e7-916d-32f09813772c.png)

**Comment by bluenevus** [comment]
your redirect url on o365 will need to be https://your.rocketchat-server.com/_oauth/CustomOAuthNAME

doing a localhost will be problematic.  If you want to do that, put a reverse proxy so you are going out and in on https on 443 not 3000

It works...we have it working this way

**Comment by geekgonecrazy** [comment]
We definitely need to turn this in to a doc.  I think this would be incredibly useful.

**Comment by JannikZed** [comment]
I don't know what I'm doing wrong, but it's still not working for me. I'm using exactly the settings you where screenshotting, but no success. It would be so nice to have o365 as an officially tested oauth provider in rocketchat.
		
		

		TRIEB.WORK | DISRUPTIVE IT SOLUTIONS
Jannik Zinkl
Geschäftsleitung
+49 151 51183272 (mobil)
Email: jannik.zinkl@trieb.work

Diese E-Mail ist ausschließlich für den angeführten Adressaten bestimmt. Sie kann Informationen enthalten, welchevertraulich sind und durch die Verschwiegenheit geschützt werden. „trieb.work“ und „BitPaket“ sind Markennamen dertrieb.work Zinkl, Marquart GbR, Vorrastraße 9 90482 Nürnberg. USt.ID.: DE310650608, Kontakt: info@trieb.work
	




On Sat, Jul 22, 2017 at 4:57 AM +0200, "Aaron Ogle" <notifications@github.com> wrote:












We definitely need to turn this in to a doc.  I think this would be incredibly useful.

**Comment by bluenevus** [comment]
The problem is you are in localhost on http vs a domain or up on https port 443

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by KoAi** [comment]
Still not working for me too with same config and https with domain.
I will give others tries next week

**Comment by JannikZed** [comment]
Hi,
@bluenevus that's not right. I'm a different person :D we are using https on a FQDN. This is what my config looks like:
![bildschirmfoto 2017-07-29 um 01 37 58](https://user-images.githubusercontent.com/7526174/28739842-c706b60c-73fe-11e7-927f-8edc3f643bdb.png)
![bildschirmfoto 2017-07-29 um 01 37 43](https://user-images.githubusercontent.com/7526174/28739843-c72f39b0-73fe-11e7-8ae7-c530306fabe1.png)

And this is what I get in my log:
![bildschirmfoto 2017-07-29 um 01 39 04](https://user-images.githubusercontent.com/7526174/28739849-d3a1f87c-73fe-11e7-9449-6015bc737ff7.png)

**Comment by steversk** [comment]
I have followed the steps above and can successfully authenticate but it fails when trying to return to the redirect URL. After authenticating, the authentication screen hangs while trying to return back to the redirect URL listed at apps.microsoft.com. 

I'm setting the redirect URL to the same as the callback URL listed in my custom OAuth:
     https://rapchat-dev.corp.*****.net/_oauth/azuread

When I try to connect to the Callback URL listed in RocketChat, it works fine until I add an entry to the URL field. Once, I add https://login.microsoftonline.com/common to the URL field, I can no longer connect to the Callback URL and instead receive a timeout error.

I've attached a screenshot of the configuration in RocketChat. I've set the redirect URL at apps.microsoft.com to match the callback URL.

![ss](https://user-images.githubusercontent.com/17030035/30286370-a0758e84-96de-11e7-8e8d-d03f91ad16cb.JPG)


Thanks for your help!

**Comment by bluenevus** [comment]
do you have all of the other values like this image?


Your callback url that you setup in office365 comes from this statement here on the top of the oauth configuration screen below


"When setting up your OAuth Provider, you'll have to inform a Callback URL. Use..."


It has to be HTTPS


--
Dr. Frank Tucker
CEO
MicroHealth, LLC
People * Value * Innovation

O: (855) 294-3547 |  F: (855)-850-1264
Address:  8245 Boone Blvd, Suite 706  |  Vienna, VA 22182
<http://www.microhealthonline.com/>Email: frank.tucker@microhealthllc.com<mailto:frank.tucker@microhealthllc.com>
www.microhealthllc.com<http://www.microhealthllc.com/>

www.microhealthllc.com<http://www.microhealthllc.com/>
________________________________

**Comment by steversk** [comment]
Yes, we are using HTTPS and I believe I've matched all of the entries. I've pasted screenshots of my RocketChat setup, apps.microsoft.com setup screen, as well as the gateway timeout that occurs after I successfully authenticate.

![rc setup](https://user-images.githubusercontent.com/17030035/30287451-5df002de-96e2-11e7-8af1-412cdf1f794f.JPG)

![apps ms setup](https://user-images.githubusercontent.com/17030035/30287457-6330cd64-96e2-11e7-817b-4af78133940c.JPG)

![timeout](https://user-images.githubusercontent.com/17030035/30287464-688ca4ea-96e2-11e7-9fa1-99455ea64131.JPG)

**Comment by jmoont** [comment]
It seems that Office365 is set to use the body (payload) for the authorize and token and then the header for subsequent requests. Could you add the option to set "Token Sent Via" to something different for the identity request vs the Authorize and Token requests?

**Comment by rodrigok** [comment]
@jmoont We can try 😄  @mrinaldhar will start working on OAuth improvements soon, he will be able to answer this question soon.

**Comment by jmoont** [comment]
That sounds great but I'd like to confirm my idea (and get it working sooner!) - I've got RocketChat running on docker, on Ubuntu, on AWS - is there a way to build a version and deploy to the container or edit/hack the code on the container? Thanks.

**Comment by jmoont** [comment]
Ok - I worked it out - edited/hacked it on my docker container and its working :)

**Comment by rodrigok** [comment]
@jmoont Can you show me what exactly you did? So I can try to fix it ASAP

**Comment by jmoont** [comment]
So I just commented out the case statement for the identity so that it used the header and then set my custom oauth to use the payload for the initial requests. This is obviously a hack to get it working but having the option to use a different "Token Sent Via" for the identity request vs the authorize and token ones would be a generic fix.

**Comment by rodrigok** [comment]
@jmoont can you check if this PR solves your problem https://github.com/RocketChat/Rocket.Chat/pull/9034 ?

**Comment by jmoont** [comment]
Yes - works well. Thank you for the quick turnaround.

**Comment by rodrigok** [comment]
Closed by https://github.com/RocketChat/Rocket.Chat/pull/9034

**Comment by rodrigok** [comment]
Awesome, thanks @jmoont

**Comment by JannikZed** [comment]
@jmoont Hi, just one more little question: do you also know what to put into "Username field"? Right now rocket.chat is not recognizing the email Adresse of the authenticating user.
Kind regards

**Comment by jszaszvari** [comment]
@JannikZed 

userPrinciplename

**Comment by alincalinciuc** [comment]
@JannikZed if you need to setup email as Username you can use `upn` or any of the fields of the JWT token found here : https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-protocols-oauth-code#jwt-token-claims

**Comment by sketch242** [comment]
@alincalinciuc I used upn, but according to ticket #10196 @ is not an allowed character in usernames.  

I did find that you can also use "name", which returns the user's full name with a space, however, then mentions are broken (see ticket #7280).

I figured I would mention it here in case anyone didn't realize that you can actually leave the username field blank, and RocketChat will prompt the user at initial login.

It would be nice if there was some way to automatically set a consistent username, with functional mentions, though.  Is there any way to combine multiple fields?  I tried a few variations like {given_name}.{last_name} but couldn't find anything which worked.

**Comment by BlackFenix2** [comment]
i referred this constraint into a feature request here: #11647

**Comment by oformoso1974** [comment]
Could be knid of late, but i leave to you my doc after days of dealing with all problems.
I have Rocket.Chat authorizing against Azure AD using Oauth.
Hope to enjoy it!
[Configuracion SSO Azure AD Rocket Chat (Oauth).docx]
[Configuracion SSO Azure AD Rocket Chat (Oauth).pdf](https://github.com/RocketChat/Rocket.Chat/files/6360468/Configuracion.SSO.Azure.AD.Rocket.Chat.Oauth.pdf)

**Comment by Leo0liveira** [comment]
Can I login to rocket.chat using AzureAd with this tutorial?

**Comment by oformoso1974** [comment]
Thats exactly the idea...

**Comment by Leo0liveira** [comment]
After having done the configuration (which in the latest versions is different) what should i do? Sorry i'm learning how to use this now.

**Comment by Leo0liveira** [comment]
It is done but when i try to login i got back to login page
This is the logs and the config of Custom Oauth:
![logs](https://user-images.githubusercontent.com/54418482/171860092-b3144544-130c-4efc-8fa8-59b2b707c37d.png)

**Comment by oformoso1974** [comment]
You press the "Azure AD Login" button to login?
Im not currently have rocket chat running in my Company so I have not all the details.

**Comment by Leo0liveira** [comment]
Yeah i pressed it generates the popup screen i select my login and i am redirected to the first step again which would be to choose the account I want to login to

**Comment by oformoso1974** [comment]
I dont think so, I was using freeversion of both, too.
Could be something wrong with the login field. The process should send you to a Azure window to validate it against it

**Comment by Leo0liveira** [comment]
Maybe i need to change the UPN options on login and email fields?

**Comment by howdoiworkthis** [comment]
> It is done but when i try to login i got back to login page This is the logs and the config of Custom Oauth: ![logs](https://user-images.githubusercontent.com/54418482/171860092-b3144544-130c-4efc-8fa8-59b2b707c37d.png) Can someone help me?

Hi all,
I found this thread very useful in getting my own configuration to work, thanks to everyone.  The settings in this image are all good except for a few changes which finally worked for me.

Merge users from distinct services: off (this turned out to be the magic bullet.  When on, it would succeed, but would prompt to make a new user instead of merging with existing.)
URL: https://login.microsoftonline.com/<client_id>
Scope: openid
Key Field: Email
Email field: upn (this is the Email address returned from Azure (now Entra))
Merge users: on

I also had to add an additional callback URL: https://chat.eurotrol.com/_oauth/o365azure?close
because the one specified in the configuration failed with an actually helpful error message.
